// typeorm/decorator/columns/PrimaryGeneratedColumn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../../index");
/**
 * Column decorator is used to mark a specific class property as a table column.
 * Only properties decorated with this decorator will be persisted to the database when entity be saved.
 * This column creates an integer PRIMARY COLUMN with generated set to true.
 * This column creates is an alias for @PrimaryColumn("int", { generated: true }).
 */
function PrimaryGeneratedColumn(strategyOrOptions, maybeOptions) {
    var options = {};
    if (strategyOrOptions) {
        if (typeof strategyOrOptions === "string")
            options.generationStrategy = strategyOrOptions;
        if (strategyOrOptions instanceof Object)
            Object.assign(options, strategyOrOptions);
    }
    else {
        options.generationStrategy = "increment";
    }
    if (maybeOptions instanceof Object)
        Object.assign(options, maybeOptions);
    return function (object, propertyName) {
        // check if there is no type in column options then set the int type - by default for auto generated column
        if (!options.type) {
            if (options.generationStrategy === "increment") {
                Object.assign(options, { type: Number });
            }
            else {
                Object.assign(options, { type: "uuid" });
            }
        }
        // implicitly set a primary and generated to column options
        Object.assign(options, { primary: true, generated: true });
        // create and register a new column metadata
        var args = {
            target: object.constructor,
            propertyName: propertyName,
            mode: "regular",
            options: options
        };
        index_1.getMetadataArgsStorage().columns.push(args);
    };
}
exports.PrimaryGeneratedColumn = PrimaryGeneratedColumn;
